text/0000-pi-types.md

```diff
@@ -0,0 +1,1139 @@
+upper and lower bound, checked at compile time):
+
+```rust
+use std::ops;
+
+/// A bounded integer.
+///
+/// This has two value parameter, respectively representing an upper and a lower bound.
+pub struct BoundedInt<const lower: usize, const upper: usize>
+    // Compile time constraints.
+    where lower <= upper {
+    /// The inner runtime value.
+    n: usize,
+}
+
+// To see how this holds the `where` clause above, see the section on `identities`.
+impl<const n: usize> BoundedInt<n, n> {
+    fn new() -> Self {
+        BoundedInt {
+            n: n,
+        }
+    }
+}
+
+/// Addition of two `BoundedInt` will simply add their bounds.
+///
+/// We check for overflow making it statically overflow-free calculations.
+impl<const upper_a: usize,
+     const lower_a: usize,
+     const upper_b: usize,
+     const lower_b: usize> ops::Add<BoundedInt<lower_b, upper_b>> for BoundedInt<lower_a, upper_a>
+     // We have to satisfy the constraints set out in the struct definition.
+     where lower_a <= upper_a,
+           lower_b <= upper_b,
+           // Check for overflow by some `const fn`.
+           is_overflow_safe(upper_a, upper_b) {
+    // These parameters are constant expression.
+    type Output = BoundedInt<lower_a + lower_b, upper_a + upper_b>;
+
+    fn add(self, rhs: BoundedInt<lower_b, upper_b>) -> Self::Output {
+        BoundedInt {
+            n: self.n + rhs.n,
+        }
+    }
+}
+
+impl<const upper_a: usize,
+     const lower_a: usize,
+     const upper_b: usize,
+     const lower_b: usize> From<BoundedInt<lower_b, upper_b>> for BoundedInt<lower_a, upper_a>
+     where lower_a <= upper_a,
+           lower_b <= upper_b,
+           // We will only extend the bound, never shrink it without runtime
+           // checks, thus we add this clause:
+           lower_b <= lower_a && upper_b >= upper_a {
+    fn from(from: BoundedInt<lower_b, upper_b>) -> Self {
+        BoundedInt {
+            n: from.n,
+        }
+    }
+}
+```
+
+### Homogeneous varargs
+
+We can use arbitrarily length arrays to simulate homogeneous varargs:
+
+```rust
+fn my_func<const n: usize>(args: [u32; n]) { /* whatever */ }
+
+my_func([1, 2, 3]);
+my_func([1, 2, 3, 4]);
+```
+
+### Array generics
+
+Currently libcore only implements various traits up to arrays of length 32.
+This allows for implementing them for arrays of arbitrary length:
+
+```rust
+impl<const n: usize, T: Clone> Clone for [T; n] {
+    fn clone(&self) -> [T; n] {
+        // Clone it...
+    }
+}
+```
+
+### Statically checked indexing
+
+One can perform simple, interval based, statically checked indexing:
+
+```rust
+use std::ops;
+
+impl<const n: usize, T: Clone> ops::Index<BoundedInt<0, n - 1>> for [T; n] {
+    type Output = T;
+
+    fn index(&self, ind: BoundedInt<0, n - 1>) -> &T {
+        unsafe {
+            // This is safe due to the bound on `ind`.
+            self.unchecked_index(*ind)
+        }
+    }
+}
+```
+
```

Comment:
one example that I personally hope is possible with this RFC:

```rust
use std::ops;

struct ConstantRange<const lo: usize, const hi: usize> where lo < hi;

impl<const lo: usize, const hi: usize, const n: usize, T: Clone> ops::Index<ConstantRange<lo, hi>> for [T; n]
    where lo < hi,
        hi <= n,
{
    type Output = [T; hi - lo];

    // index value is irrelevant, since it's a unit struct
    fn index(&self, _: ConstantRange<lo, hi>) -> &[T; hi - lo] {
        unsafe {
            // This is safe due to the bounds on `lo` and `hi`.
            &*(self.get_unchecked(lo) as *const T as *const [T; hi - lo])
        }
    }
}

// call it like
let five_element_array_view = &some_array[ConstantRange::<5, 10>];
```

or for slices:

```rust
use std::ops;

struct ConstantRange<const lo: usize, const hi: usize> where lo < hi;

impl<const lo: usize, const hi: usize, T: Clone> ops::Index<ConstantRange<lo, hi>> for [T]
    where lo < hi,
{
    type Output = [T; hi - lo];

    // index value is irrelevant, since it's a unit struct
    fn index(&self, _: ConstantRange<lo, hi>) -> &[T; hi - lo] {
        unsafe {
            // panics at runtime if out of range
            &*(&self[lo] as *const T as *const [T; hi - lo])
        }
    }
}

// call it like
let five_element_array_view = &some_slice[ConstantRange::<5, 10>];
```